hugr-core/src/extension/resolution/types_mut.rs

```diff
@@ -0,0 +1,212 @@
+        }
+        OpType::FuncDefn(f) => {
+            update_signature_exts(node, f.signature.body_mut(), extensions, used_extensions)?
+        }
+        OpType::FuncDecl(f) => {
+            update_signature_exts(node, f.signature.body_mut(), extensions, used_extensions)?
+        }
+        OpType::Const(c) => {
+            let typ = c.get_type();
+            let mut missing = HashSet::new();
+            collect_type_exts(&typ, used_extensions, &mut missing);
+            // We expect that the `CustomConst::get_type` binary calls always return valid extensions.
+            // As we cannot update the `CustomConst` type, we ignore the result.
+            //
+            // Some exotic consts may need https://github.com/CQCL/hugr/issues/1742 to be implemented
+            // to pass this test.
+            //assert!(missing.is_empty());
+        }
+        OpType::Input(inp) => {
+            update_type_row_exts(node, &mut inp.types, extensions, used_extensions)?
+        }
+        OpType::Output(out) => {
+            update_type_row_exts(node, &mut out.types, extensions, used_extensions)?
+        }
+        OpType::Call(c) => {
+            update_signature_exts(node, c.func_sig.body_mut(), extensions, used_extensions)?;
+            update_signature_exts(node, &mut c.instantiation, extensions, used_extensions)?;
+        }
+        OpType::CallIndirect(c) => {
+            update_signature_exts(node, &mut c.signature, extensions, used_extensions)?
+        }
+        OpType::LoadConstant(lc) => {
+            update_type_exts(node, &mut lc.datatype, extensions, used_extensions)?
+        }
+        OpType::LoadFunction(lf) => {
+            update_signature_exts(node, lf.func_sig.body_mut(), extensions, used_extensions)?;
+            update_signature_exts(node, &mut lf.signature, extensions, used_extensions)?;
+        }
+        OpType::DFG(dfg) => {
+            update_signature_exts(node, &mut dfg.signature, extensions, used_extensions)?
+        }
+        OpType::OpaqueOp(op) => {
+            update_signature_exts(node, op.signature_mut(), extensions, used_extensions)?
+        }
+        OpType::Tag(t) => {
+            for variant in t.variants.iter_mut() {
+                update_type_row_exts(node, variant, extensions, used_extensions)?
+            }
+        }
+        OpType::DataflowBlock(db) => {
+            update_type_row_exts(node, &mut db.inputs, extensions, used_extensions)?;
+            update_type_row_exts(node, &mut db.other_outputs, extensions, used_extensions)?;
+            for row in db.sum_rows.iter_mut() {
+                update_type_row_exts(node, row, extensions, used_extensions)?;
+            }
+        }
+        OpType::ExitBlock(e) => {
+            update_type_row_exts(node, &mut e.cfg_outputs, extensions, used_extensions)?;
+        }
+        OpType::TailLoop(tl) => {
+            update_type_row_exts(node, &mut tl.just_inputs, extensions, used_extensions)?;
```

Comment:
Hmmm, so the trick with just scanning the signature doesn't work here. I'm tempted to suggest a method on OpType `fn all_types_mut(&mut self) -> Iterator<Item = &mut Type>` but not sure that'd be an improvement (it's probably much like this, in contrast to using Signature that exists already)

Reply:
That's too specific to what we are doing here. This is an internal method that fiddles with the internal, `all_types_mut` wouldn't have much use outside of it.